File: ruby/solutions/solution1.rb
# frozen_string_literal: true

require_relative 'solution'

# Solution to December 1
class Solution1 < Solution

  private

  def first_solution(lines)
    lines.map(&:to_i).each_cons(2).select { |x| x[1] > x[0] }.count
  end

  def second_solution(lines)
    lines.map(&:to_i).each_cons(4).select { |x| x[3] > x[0] }.count
  end
end

File: dotnet/AdventOfCode/Solutions/Solution1.cs
using System.Collections.Generic;
using System.Linq;

namespace AdventOfCode.Solutions
{
    public class Solution1 : Solution
    {
        protected override int FirstSolution(List<string> lines) => GeneralSolution(lines, 1);

        protected override int SecondSolution(List<string> lines) => GeneralSolution(lines, 3);

        private static int GeneralSolution(IReadOnlyCollection<string> lines, int indexesBack) => 
            ConvertToIntegers(lines).Where((v, i) => i >= indexesBack && v > ConvertToIntegers(lines)[i - indexesBack]).Count();
    }
}